test/scraper/scraper.service.spec.ts

```diff
@@ -1,26 +1,142 @@
 import { Test, TestingModule } from '@nestjs/testing';
 import { ScraperService } from 'src/modules/scraper/scraper.service';
 
-describe('Scraper Service', () => {
-  let scraperService: ScraperService;
-  let fsWriteFileSyncSpy: jest.SpyInstance;
-  let fsWriteFileSyncMock: jest.Mock;
-  beforeEach(async () => {
+describe('ScraperService', () => {
+  let service: ScraperService;
+
+  beforeAll(async () => {
     const module: TestingModule = await Test.createTestingModule({
-      providers: [
-        ScraperService,
-        {
-          provide: 'PUPPETEER',
-          useValue: {
-            launch: jest.fn(),
-          },
-        },
-      ],
+      providers: [ScraperService],
     }).compile();
-    scraperService = module.get<ScraperService>(ScraperService);
-    fsWriteFileSyncMock = jest.fn();
+
+    service = module.get<ScraperService>(ScraperService);
+    await service.init(); // Initialize the browser instance
+  }, 5000);
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
+  });
+  describe('fetchPage', () => {
+    it('should fetch page content', async () => {
+      const query = 'test';
+      const start = 0;
+      const pageContent = await service.fetchPage(query, start);
+
+      // Ensure that page content is not empty
+      expect(pageContent).toBeTruthy();
+    });
+  });
+  describe('parseResults', () => {
+    it('should parse search results from HTML', () => {
+      const html = `
+      <div>
+        <span>Được tài trợ</span>
+        <div id="result-stats">Total search results: 100</div>
+        <div id="rso">
+          <a href="https://fb.com">
+            <h3>Title 1</h3>
+          </a>
+          <a href="https://fb.com">
+            <h3>Title 2</h3>
+          </a>
+        </div>
+      </div>
+    `;
+
+      const parsedResults = service.parseResults(html);
+      // Ensure that search results are parsed correctly
+      expect(parsedResults.totalAds).toBe(1);
+      expect(parsedResults.totalLinkOfPage).toBe(2);
+      expect(parsedResults.searchResults).toEqual([
+        { title: 'Title 1', link: 'https://fb.com' }, // You need to adjust this based on your HTML structure
+        { title: 'Title 2', link: 'https://fb.com' }, // You need to adjust this based on your HTML structure
+      ]);
+      expect(parsedResults.totalSearchResult).toBe('Total search results: 100');
+    });
+
+    it('should return empty search results if no results found', () => {
+      const html = `
+      <div>
+        <span>Được tài trợ</span>
+          <div id="result-stats">Total search results: 0</div>
+        <div id="rso">
+        </div>
+      </div>
+    `;
+
+      const parsedResults = service.parseResults(html);
+      // Ensure that search results are empty
+      expect(parsedResults.searchResults).toEqual([]);
+      expect(parsedResults.totalSearchResult).toBe('Total search results: 0');
+    });
+
+    it('should return totalAds base on Được tài trợ', () => {
```

Comment:
```suggestion
    describe('given 2 HTML elements with the text content "Được tài trợ"', () => {
      it('should return 2 totalAds', () => {
```

I recommend using another "block" to isolate each test context (i.e. 1 or many tests that use the exact same input data). This helps differentiate the "context" (state of data) from the "expectation" (what to assert). :)